File: coupon-common/src/main/java/com/kakao/mobility/kmcoupon/domain/coupon/Coupon.java
package com.kakao.mobility.kmcoupon.domain.coupon;

import javax.persistence.*;
import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "coupon")
public class Coupon {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    private BigDecimal useMinAmount;

    private BigDecimal discountAmount;

    private LocalDateTime usableFrom;

    private LocalDateTime usableUntil;

    @Enumerated(EnumType.STRING)
    private Status status;

    private Long memberId;

    public Coupon() {
    }

    public Coupon(
            BigDecimal useMinAmount,
            BigDecimal discountAmount,
            LocalDateTime usableFrom,
            LocalDateTime usableUntil,
            Long memberId) {
        this.useMinAmount = useMinAmount;
        this.discountAmount = discountAmount;
        this.usableFrom = usableFrom;
        this.usableUntil = usableUntil;
        this.memberId = memberId;

        this.status = Status.NORMAL;
    }

    public Long getId() {
        return id;
    }

    public BigDecimal getUseMinAmount() {
        return useMinAmount;
    }

    public BigDecimal getDiscountAmount() {
        return discountAmount;
    }

    public LocalDateTime getUsableFrom() {
        return usableFrom;
    }

    public LocalDateTime getUsableUntil() {
        return usableUntil;
    }

    public Status getStatus() {
        return status;
    }

    public Long getMemberId() {
        return memberId;
    }

    public void used() {
        this.status = Status.USED;
    }

    public boolean isAlreadyUsed() {
        return this.status == Status.USED;
    }

    public boolean isMinAmountBiggerThan(BigDecimal itemAmount) {
        return this.useMinAmount.compareTo(itemAmount) > 0;
    }

    public boolean isNotUsableDuration(LocalDateTime requestReceivedTime) {
        return this.usableFrom.isAfter(requestReceivedTime)
                || this.usableUntil.isBefore(requestReceivedTime);
    }
}

File: coupon-common/src/main/java/com/kakao/mobility/kmcoupon/domain/coupon/Coupon.kt
package com.kakao.mobility.kmcoupon.domain.coupon

import com.kakao.mobility.kmcoupon.domain.BaseTimeEntity
import java.math.BigDecimal
import java.time.LocalDateTime
import javax.persistence.*

@Entity
@Table(name = "coupon")
class Coupon(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0L,

    @Column(name = "use_min_amount")
    val useMinAmount: BigDecimal,

    @Column(name = "discount_amount")
    val discountAmount: BigDecimal,

    @Column(name = "usable_from")
    val usableFrom: LocalDateTime,

    @Column(name = "usable_until")
    val usableUntil: LocalDateTime,

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    var status: Status = Status.NORMAL,

    // var val
    @Column(name = "member_id")
    var memberId: Long = 0L
) : BaseTimeEntity() {
    fun isAlreadyUsed(): Boolean {
        return status == Status.USED
    }

    fun used() {
        status = Status.USED
    }

    // kotlin에서 is를 대체하는 경우가 존재
    // 0, 0.0 같은 값인지 (compareTo으로 웬만하면)
    fun isMinAmountBiggerThan(itemAmount: BigDecimal): Boolean {
        return useMinAmount > itemAmount
    }

    fun isNotUsableDuration(requestReceivedTime: LocalDateTime): Boolean {
        return (usableFrom.isAfter(requestReceivedTime)
                || usableUntil.isBefore(requestReceivedTime))
    }

    fun calPayAmount(itemAmount: BigDecimal): BigDecimal {
        // 이부분도
        return if (itemAmount <= discountAmount) {
            BigDecimal.ZERO
        } else {
            itemAmount.subtract(discountAmount)
        }
    }

    fun calActualDiscountAmount(itemAmount: BigDecimal): BigDecimal {
        return if (itemAmount <= discountAmount) {
            itemAmount
        } else {
            discountAmount
        }
    }
}